### ISSUE TITLE
Windows: arbitrary linking of libraries

### ISSUE THREAD

**Issue by MarnixKuijs** [rust-lang/rust#80126]
In `library/std/src/sys/windows/mod.rs` some libraries, for either uwp or win32, get linked. The block that follows these linking statements contain no declarations, so to me it seems that these libraries are just linked for the sake of linking them, but it doesn't look like anything depends on them.

```rust
cfg_if::cfg_if! {
    if #[cfg(target_vendor = "uwp")] {
        #[link(name = "ws2_32")]
        // For BCryptGenRandom
        #[link(name = "bcrypt")]
        extern "C" {}
    } else {
        #[link(name = "advapi32")]
        #[link(name = "ws2_32")]
        #[link(name = "userenv")]
        extern "C" {}
    }
}
```  

Even when `no-default-libraries` is set to true in a target config they are still getting linked. This is very unfortunate, because I have to make something that relies on `/NODEFAULTLIB` to exclude the arbitrary linking of these libraries. The [commit](https://github.com/rust-lang/rust/commit/6f3872a14c73a74c55e0854641f4dfe5e3b7000e) that made this change was to clean up the build script. The build script where they got originally linked also didn't have a way to opt-out of them, but I would like to have a method to opt-out of the linking of these libs, so that I can provide my own set of windows libraries.

**Comment by MarnixKuijs** [comment]
Now looking at it closer it does it for basically every platform, so it might be nice if we can come up with a way to be able to opt-out of linking these libs for all platform so that people can provide there own.

**Comment by sfackler** [comment]
They are linked because the standard library depends on functions they provide. As the comment says in the code you copied in, we depend on bcrypt for `BCryptGenRandom`.

**Comment by MarnixKuijs** [comment]
Right, my first comment was maybe a bit malformed. My second comment is more the direction that I want to get at. I want to provide my own system libs, in my case windows system libs, that provide the same functionality as the libs that are always linked right now. I would like to see, for example, a feature `external_system_libs`, which will disable the linking of the libs in std. This should be fine as long as the user will link the system libs else where for example in there executable.

**Comment by RodBurman** [comment]
Maybe I have this wrong but doesn't #![no_std] achieve what is being requested? Since this is used for Kernel development where the standard libraries are neither available nor desired it may remove  more than hoped for but UI am not sure how you could "have your cake and eat it", i.e. use the standard libraries without their underpinnings at least from a language point of view, if the expected libraries are removed then the substitutes must be added manually in the build scripts.
